sdb/commands/member.py

```diff
@@ -31,16 +32,145 @@ class Member(sdb.Command):
+        terms = []
+        while tokens:
+
+            is_index = False
+            identifier = ""
+            idx = ""
+            if tokens[0] == "." or tokens[0] == "->":
+                sep = tokens[0]
+                identifier = tokens[1]
+                tokens = tokens[2:]
+            elif tokens[0] == '[':
+                if len(tokens) < 3 or tokens[2] != ']':
+                    raise sdb.CommandError(self.name,
+                                           "incomplete array expression")
+                if not tokens[1].isdigit():
+                    raise sdb.CommandError(
+                        self.name,
+                        "incorrect index: '{}' is not a number".format(
+                            tokens[1]))
+                is_index = True
+                idx = tokens[1]
+                sep = "[]"
+                tokens = tokens[3:]
+            else:
+                sep = ""
+                identifier = tokens[0]
+                tokens = tokens[1:]
+
+            if not is_index and not identifier.isidentifier():
+                raise sdb.CommandError(
+                    self.name,
+                    "{} is not an acceptable identifier".format(identifier))
+
+            if not is_index:
+                assert idx == ""
+                terms.append((sep, identifier))
+            else:
+                assert identifier == ""
+                terms.append((sep, idx))
+        return terms
+
+    @staticmethod
+    def _typedef_to_base_kind(type_: drgn.Type) -> drgn.TypeKind:
+        while type_.kind == drgn.TypeKind.TYPEDEF:
+            type_ = type_.type
+        return type_.kind
+
+    def _validate_type_dereference(self, obj: drgn.Object, sep: str) -> None:
+        kind = self._typedef_to_base_kind(obj.type_)
+
+        # This is the first term, no need to do validation.
+        if not sep:
+            return
+
+        if kind == drgn.TypeKind.STRUCT and sep != ".":
+            raise sdb.CommandError(
+                self.name,
+                "'{}' is a struct - use the dot(.) notation for member access".
+                format(obj.type_))
+
+        if kind == drgn.TypeKind.POINTER and sep != "->" and sep != '.':
+            raise sdb.CommandError(
+                self.name,
+                "'{}' is a pointer - use the arrow(->) notation for member access"
+                .format(obj.type_))
+
+        if kind == drgn.TypeKind.ARRAY and sep != "[]":
+            raise sdb.CommandError(
+                self.name, "'{}' is an array - cannot use '{}' notation".format(
+                    obj.type_, sep))
+
+    def _validate_array_index(self, type_: drgn.Type, idx: int) -> None:
+        assert Member._typedef_to_base_kind(type_) == drgn.TypeKind.ARRAY
+        if type_.length <= idx:
+            raise sdb.CommandError(
+                self.name,
+                "index out of bounds for array of type '{}' (requested index: {})"
+                .format(type_, idx))
```

Comment:
After giving this some thought I decided that this error is more useful as a warning that is printed and we do the dereference anyway - instead of throwing an error and stopping everything just because of zero-length arrays like the one you mentioned above.